_posts/2020-11-10-jit-part-2.md

```diff
@@ -0,0 +1,192 @@
+    Label after_gc_check = a.newLabel();
+
+    /* Do we have enough free space already? */
+    a.lea(ARG2, x86::qword_ptr(HTOP, words_needed * sizeof(Eterm)));
+    a.cmp(ARG2, E);
+    a.jbe(after_gc_check);
+
+    /* No, we need to GC.
+     *
+     * Switch to the C stack, and update the process
+     * structure with our current stack (E) and heap
+     * (HTOP) pointers so the C code can use them. */
+    emit_enter_runtime<Update::eStack | Update::eHeap>();
+
+    /* Call the GC, passing how many words we need and
+     * how many X registers we use. */
+    a.mov(ARG2, imm(words_needed));
+    a.mov(ARG4, imm(Live.getValue()));
+    a.mov(ARG1, c_p);
+    load_x_reg_array(ARG3);
+    a.mov(ARG5, FCALLS);
+    runtime_call<5>(erts_garbage_collect_nobump);
+    a.sub(FCALLS, RET);
+
+    /* Swap back to the Erlang stack, reading the new
+     * values for E and HTOP from the process structure. */
+    emit_leave_runtime<Update::eStack | Update::eHeap>();
+
+    a.bind(after_gc_check);
+}
+```
+
+While this isn't too complicated it's still a rather large amount of code:
+since all instructions are emitted directly into their modules, small
+inefficiencies like this tend to bloat the modules rather quickly. Beyond using
+more memory, this wastes precious instruction cache so we've spent a lot of
+time and effort on reducing code size.
+
+Our most common method of reducing code size is to break out as much of the
+instruction as possible into a globally shared part. Let's see how we can apply
+this technique:
+
+```c++
+void BeamModuleAssembler::emit_test_heap(const ArgVal &Needed,
+                                         const ArgVal &Live) {
+    const int words_needed = (Needed.getValue() + S_RESERVED);
+    Label after_gc_check = a.newLabel();
+
+    a.lea(ARG2, x86::qword_ptr(HTOP, words_needed * sizeof(Eterm)));
+    a.cmp(ARG2, E);
+    a.jbe(after_gc_check);
+
+    a.mov(ARG4, imm(Live.getValue()));
+
+    /* Call the global "garbage collect" fragment. */
+    fragment_call(ga->get_garbage_collect());
+
+    a.bind(after_gc_check);
+}
+
+/* ARG2 = (HTOP + bytes needed) !!
+ * ARG4 = Live registers */
+void BeamGlobalAssembler::emit_garbage_collect() {
+    /* As ARG2 is (HTOP + bytes needed), we can convert
+     * it to "words needed" by subtracting HTOP and
+     * dividing it by 8.
+     *
+     * This saves us from having to explicitly pass
+     * "words needed" in the module code above. */
+    a.sub(ARG2, HTOP);
+    a.shr(ARG2, imm(3));
+
+    emit_enter_runtime<Update::eStack | Update::eHeap>();
+
+    /* ARG2 and ARG4 have already been set earlier. */
+    a.mov(ARG1, c_p);
+    load_x_reg_array(ARG3);
+    a.mov(ARG5, FCALLS);
+    runtime_call<5>(erts_garbage_collect_nobump);
+    a.sub(FCALLS, RET);
+
+    emit_leave_runtime<Update::eStack | Update::eHeap>();
+
+    a.ret();
+}
+```
+
+While we had to write about as much code, the part that is copied into the
```

Comment:
I think that it would make sense to mention that we use our own calling convention here and not System V.